src/gfn/envs/box.py

```diff
@@ -91,12 +91,13 @@ def is_action_valid(
             if torch.any(self.norm(actions_at_s0) > self.delta):
                 return False
 
-        non_s0_states = non_terminal_states[~s0_states_idx].states_tensor
+        non_s0_states = non_terminal_states[~s0_states_idx].tensor
         non_s0_actions = non_exit_actions[~s0_states_idx].tensor
 
-        if torch.any(self.norm(non_s0_actions) != self.delta) or torch.any(
-            non_s0_actions < 0
-        ):
+        if (
+            not backward
+            and torch.any(torch.abs(self.norm(non_s0_actions) - self.delta) > 1e-5)
```

Comment:
magic number! would prefer this be defined or set as a `tol=` parameter in the env.